File: src/main/java/slavsquatsuperstar/mayonez/physics2d/colliders/PolygonCollider2D.java
package slavsquatsuperstar.mayonez.physics2d.colliders;

import org.apache.commons.lang3.ArrayUtils;
import slavsquatsuperstar.math.Mat22;
import slavsquatsuperstar.math.MathUtils;
import slavsquatsuperstar.math.Vec2;
import slavsquatsuperstar.mayonez.physics2d.CollisionManifold;
import slavsquatsuperstar.mayonez.physics2d.RaycastResult;

public abstract class PolygonCollider2D extends Collider2D {

    private final Vec2[] vertices;

    // points in local space (relative to center)

    /**
     * Constructs a convex polygon from an array of vertices in clockwise order.
     *
     * @param vertices the vertices in clockwise order
     */
    public PolygonCollider2D(Vec2... vertices) {
        this.vertices = vertices;
    }

    // Shape Properties

    public float getRotation() {
        return transform.rotation;
    }

    public final int countVertices() {
        return vertices.length;
    }

    public Vec2[] getVertices() { // rotated in world
        Vec2[] results = new Vec2[countVertices()];
        for (int i = 0; i < results.length; i++)
            results[i] = toWorld(vertices[i]);
        return results;
    }

    @Override
    public AlignedBoxCollider2D getMinBounds() { // TODO Support function
        Vec2[] vertices = getVertices();
        float[] verticesX = new float[vertices.length];
        float[] verticesY = new float[vertices.length];

        // Get the min and max coordinates of any point on the box
        for (int i = 0; i < vertices.length; i++) {
            verticesX[i] = vertices[i].x;
            verticesY[i] = vertices[i].y;
        }
        Vec2 newMin = new Vec2(MathUtils.min(verticesX), MathUtils.min(verticesY));
        Vec2 newMax = new Vec2(MathUtils.max(verticesX), MathUtils.max(verticesY));

        Vec2 aabbSize = newMax.sub(newMin).div(transform.scale);
        return new AlignedBoxCollider2D(aabbSize).setTransform(transform).setRigidBody(rb);
    }

    // Intersection Helper Methods

    public Edge2D[] getEdges() { // in world
        Vec2[] vertices = getVertices();
        Edge2D[] edges = new Edge2D[vertices.length];
        for (int i = 0; i < edges.length; i++)
            edges[i] = new Edge2D(vertices[i], vertices[(i + 1) % edges.length]); // Current vertex to next vertex
        return edges;
    }

    public Vec2[] getNormals() { // in world
        Edge2D[] edges = getEdges();
        Vec2[] normals = new Vec2[countVertices()];
        Mat22 rot = new Mat22(0, 1, -1, 0); // Rotate 90 degrees counterclockwise to point outward
        for (int i = 0; i < normals.length; i++)
            normals[i] = rot.times(edges[i].toVector()).unit();
        return normals;
    }

    // Separating-Axis Theorem Methods

    // Project vertices on the axis and find min and max
    protected final MathUtils.Range getIntervalOnAxis(Vec2 axis) {
        Vec2[] vertices = getVertices();
        float[] projections = new float[countVertices()];
        for (int i = 0; i < projections.length; i++)
            projections[i] = vertices[i].dot(axis);
        return new MathUtils.Range(MathUtils.min(projections), MathUtils.max(projections));
    }

    protected final boolean overlapOnAxis(PolygonCollider2D polygon, Vec2 axis) {
        MathUtils.Range thisInterval = this.getIntervalOnAxis(axis);
        MathUtils.Range otherInterval = polygon.getIntervalOnAxis(axis);
        return (thisInterval.min <= otherInterval.max) && (otherInterval.min <= thisInterval.max);
    }

    // Assume overlap on axis
    protected final float getAxisOverlap(PolygonCollider2D polygon, Vec2 axis) {
        MathUtils.Range thisInterval = this.getIntervalOnAxis(axis);
        MathUtils.Range otherInterval = polygon.getIntervalOnAxis(axis);
        return Math.min(Math.abs(thisInterval.min - otherInterval.max), Math.abs(otherInterval.min - thisInterval.max));
    }

    // Polygon vs Point

    @Override
    public boolean contains(Vec2 point) {
        Edge2D[] edges = getEdges();
        for (Edge2D edge : edges) {
            Vec2 edgeLine = edge.toVector();
            float projLength = point.sub(edge.start).projectedLength(edgeLine);
            if (!MathUtils.inRange(projLength, 0, edge.length()))
                return false;
        }
        return true;
    }

    @Override
    public Vec2 nearestPoint(Vec2 position) {
        if (contains(position))
            return position;
        Edge2D[] edges = getEdges();
        float[] distances = new float[edges.length];
        for (int i = 0; i < distances.length; i++)
            distances[i] = position.distance(edges[i].nearestPoint(position));
        Edge2D nearestEdge = edges[MathUtils.minIndex(distances)];
        return nearestEdge.nearestPoint(position);
    }

    // Polygon vs Line

    @Override
    public RaycastResult raycast(Ray2D ray, float limit) {
        Edge2D[] edges = getEdges();
        float[] distances = new float[edges.length];
        for (int i = 0; i < distances.length; i++) {
            RaycastResult rc = edges[i].raycast(ray, limit);
            distances[i] = (rc != null) ? rc.getDistance() : Float.MAX_VALUE;
        }
        int minIndex = MathUtils.minIndex(distances);
        float minDist = distances[minIndex];
        if (minDist == Float.MAX_VALUE) // no successful raycasts
            return null;
        Vec2 normal = edges[minIndex].toVector().rotate(-90).unit();
        return new RaycastResult(ray.getPoint(minDist), normal, minDist);
    }

    // Polygon vs Shape

    @Override
    public boolean detectCollision(Collider2D collider) {
        if (collider instanceof PolygonCollider2D)
            return intersects((PolygonCollider2D) collider);
        return getCollisionInfo(collider) != null;
    }

    // Separating-Axis Theorem
    private boolean intersects(PolygonCollider2D polygon) {
        Vec2[] axes = ArrayUtils.addAll(this.getNormals(), polygon.getNormals());
        for (Vec2 axis : axes)
            if (!overlapOnAxis(polygon, axis))
                return false;
        return true;
    }

    @Override
    public CollisionManifold getCollisionInfo(Collider2D collider) {
        if (collider instanceof CircleCollider)
            return getCollisionInfo((CircleCollider) collider);
//        if (collider instanceof PolygonCollider2D)
//            return getCollisionInfo((PolygonCollider2D) collider);
        return null;
    }
    private CollisionManifold getCollisionInfo(CircleCollider circle) {
        Vec2 closestToCircle = this.nearestPoint(circle.center());
        if (!circle.contains(closestToCircle))
            return null;

        float depth = circle.radius() - closestToCircle.distance(circle.center());
        Vec2 normal = circle.center().sub(closestToCircle).unit();
        CollisionManifold result = new CollisionManifold(this, circle, normal, depth);
        result.addContact(closestToCircle.sub(normal.mul(depth)));
        return result;
    }

}

File: src/main/kotlin/slavsquatsuperstar/mayonez/physics2d/colliders/PolygonCollider2D.kt
package slavsquatsuperstar.mayonez.physics2d.colliders

import org.apache.commons.lang3.ArrayUtils
import slavsquatsuperstar.math.Mat22
import slavsquatsuperstar.math.MathUtils.inRange
import slavsquatsuperstar.math.MathUtils.max
import slavsquatsuperstar.math.MathUtils.maxIndex
import slavsquatsuperstar.math.MathUtils.min
import slavsquatsuperstar.math.MathUtils.minIndex
import slavsquatsuperstar.math.Range
import slavsquatsuperstar.math.Vec2
import slavsquatsuperstar.mayonez.Colors
import slavsquatsuperstar.mayonez.physics2d.CollisionManifold
import slavsquatsuperstar.mayonez.physics2d.RaycastResult
import slavsquatsuperstar.mayonez.renderer.DebugDraw
import kotlin.math.abs

/**
 * A convex polygon with an arbitrary number of vertices connected by straight edges.
 *
 * @author SlavSquatSuperstar
 *
 * @constructor Constructs a convex polygon from an array of vertices in clockwise order.
 *
 * @param vertices the vertices in clockwise order
 */
open class PolygonCollider2D(vararg vertices: Vec2) : Collider2D() {

    // Shape Properties

    // points in local space (relative to center)
    private val vertices: Array<Vec2> = arrayOf(*vertices)

    // rotated in world
    open fun getVertices(): Array<Vec2> = Array(vertices.size) { toWorld(vertices[it]) }

    open fun countVertices(): Int = vertices.size

    open fun getRotation(): Float = if (transform != null) transform!!.rotation else 0f

    override fun getMinBounds(): AlignedBoxCollider2D { // TODO Support function
        val vertices = getVertices()
        val verticesX = FloatArray(vertices.size)
        val verticesY = FloatArray(vertices.size)

        // Get the min and max coordinates of any point on the box
        for (i in vertices.indices) {
            verticesX[i] = vertices[i].x
            verticesY[i] = vertices[i].y
        }
        val newMin = Vec2(min(*verticesX), min(*verticesY))
        val newMax = Vec2(max(*verticesX), max(*verticesY))
        val aabbSize = newMax.sub(newMin).div(transform!!.scale)
        return AlignedBoxCollider2D(aabbSize).setTransform<Collider2D>(transform).setRigidbody(rb)
    }

    // Intersection Helper Methods

    open fun getEdges(): Array<Edge2D> { // in world
        val vertices = getVertices()
        return Array(vertices.size) { Edge2D(vertices[it], vertices[(it + 1) % vertices.size]) }
    }

    open fun getNormals(): Array<Vec2> { // in world
        val edges = getEdges()
        val rot = Mat22(0f, 1f, -1f, 0f) // Rotate 90 degrees counterclockwise to point outward
        return Array(countVertices()) { (rot * edges[it].toVector()).unit() }
    }

    open fun getDiagonals(): Array<Edge2D> {
        val center = center()
        val vertices = getVertices()
        return Array(vertices.size) { Edge2D(vertices[it], center) }
    }

    // Separating-Axis Theorem Methods

    /*
     * Projected min and max
     * Positive is in axis direction
     * (-) ---> (+)
     */
    protected fun getIntervalOnAxis(axis: Vec2): Range {
        val vertices = getVertices()
        val projections = FloatArray(vertices.size) { vertices[it].dot(axis) }
        return Range(min(*projections), max(*projections))
    }

    // Do the intervals intersect
    private fun hasOverlapOnAxis(polygon: PolygonCollider2D, axis: Vec2): Boolean {
        val thisInterval = this.getIntervalOnAxis(axis)
        val otherInterval = polygon.getIntervalOnAxis(axis)
        return (thisInterval.min <= otherInterval.max) && (otherInterval.min <= thisInterval.max)
    }

    // How much overlap on the axis
    protected fun getOverlapOnAxis(polygon: PolygonCollider2D, axis: Vec2): Float {
        val thisInterval = this.getIntervalOnAxis(axis)
        val otherInterval = polygon.getIntervalOnAxis(axis)
        if (!hasOverlapOnAxis(polygon, axis)) return 0f
        return abs(thisInterval.min - otherInterval.max).coerceAtMost(abs(otherInterval.min - thisInterval.max))
    }

    // How much separation on the axis
    // Positive means separation, and negative means overlap
    // TODO combine separation and overlap into one function (differentiate with positive and negative)
    private fun getSeparationOnAxis(polygon: PolygonCollider2D, axis: Vec2): Float {
        val normalFace = this.getIntervalOnAxis(axis).max
        val nearestVertex = polygon.getIntervalOnAxis(axis).min
        return nearestVertex - normalFace
    }


    // Polygon vs Point
    override fun contains(point: Vec2): Boolean {
        val edges = getEdges()
        for (edge in edges) {
            val edgeLine = edge.toVector()
            val projLength = point.sub(edge.start).projectedLength(edgeLine)
            if (!inRange(projLength, 0f, edge.length)) return false
        }
        return true
    }

    override fun nearestPoint(position: Vec2): Vec2? {
        if (position in this) return position
        val edges = getEdges()
        val distances = FloatArray(edges.size) { edges[it].distance(position) }
        val nearestEdge = edges[minIndex(*distances)]
        return nearestEdge.nearestPoint(position)
    }

    // Polygon vs Line
    override fun raycast(ray: Ray2D, limit: Float): RaycastResult? {
        val edges = getEdges()
        val distances = FloatArray(edges.size)
        for (i in distances.indices) {
            val rc = edges[i].raycast(ray, limit)
            distances[i] = rc?.distance ?: Float.MAX_VALUE
        }

        val minIndex = minIndex(*distances)
        val minDist = distances[minIndex]
        if (minDist == Float.MAX_VALUE) return null // no successful raycasts

        val normal = edges[minIndex].toVector().rotate(-90f).unit()
        return RaycastResult(ray.getPoint(minDist), normal, minDist)
    }

    // Polygon vs Shape

    override fun detectCollision(collider: Collider2D?): Boolean {
        return if (collider is PolygonCollider2D) detectCollision(collider)
        else getCollisionInfo(collider) != null
    }

    // Separating-Axis Theorem
    private fun detectCollision(polygon: PolygonCollider2D): Boolean {
        val axes = ArrayUtils.addAll(getNormals(), *polygon.getNormals())
        axes.forEach { axis -> if (!hasOverlapOnAxis(polygon, axis)) return false }
        return true
    }

    override fun getCollisionInfo(collider: Collider2D?): CollisionManifold? {
        return when (collider) {
            is CircleCollider -> getCollisionInfo(collider)
            is PolygonCollider2D -> getCollisionInfo(collider)
            else -> null
        }
    }

    // Polygon vs Circle: 1 contact point
    private fun getCollisionInfo(circle: CircleCollider): CollisionManifold? {
        val closestToCircle = nearestPoint(circle.center())
        if (closestToCircle!! !in circle) return null

        val depth = circle.radius() - closestToCircle.distance(circle.center())
        val normal = (circle.center() - closestToCircle).unit()
        val result = CollisionManifold(this, circle, normal, depth)
        result.addContact(closestToCircle - (normal * depth))
        return result
    }

    // Polygon vs Polygon: 1-2 contact points
    // SAT method, convex only
    private fun getCollisionInfo(poly: PolygonCollider2D): CollisionManifold? {
        // 1. Store edges and normals in structs and calculate overlaps
        val poly1 = SATPolygon(this, poly)
        val poly2 = SATPolygon(poly, this)

        // 2. Test for a separating axis
        if (poly1.minOverlap > 0 || poly2.minOverlap > 0) return null // positive separation = no intersection

        // 3. Set the polygon with lesser overlap as reference frame
        val reference: SATPolygon // the "stationary" shape
        val incident: SATPolygon // the "incoming" shape
        if (poly1.minOverlap >= poly2.minOverlap) {
            reference = poly1
            incident = poly2
        } else {
            incident = poly1
            reference = poly2
        }

        // 4. Calculate reference and incident edges
        val minOverlapIdx = reference.minIdx
//        val minIdxInc = incident.minIdx

        val overlap = reference.minOverlap
        val colNormal = reference.normals[minOverlapIdx]
        val refEdge = reference.edges[minOverlapIdx]
//        val incEdge = incident.edges[minIdxInc]

        // Get most negative dot product = edge normal facing towards collision normal the most
        // Sometimes overlap test isn't reliable
        val dotProds = FloatArray(incident.normals.size) { colNormal.dot(incident.normals[it]) }
        val minDotIdx = minIndex(*dotProds)
        val incEdge = incident.edges[minDotIdx]

        DebugDraw.drawVector(reference.collider.center(), reference.normals[minOverlapIdx], Colors.BLUE);
        DebugDraw.drawVector(incident.collider.center(), incident.normals[minDotIdx], Colors.RED);
        DebugDraw.drawLine(refEdge, Colors.BLUE)
        DebugDraw.drawLine(incEdge, Colors.RED)

        // 5. Calculate contact points
        val collision = CollisionManifold(reference.collider, incident.collider, colNormal, -overlap)
        val clippedEdge = incEdge.clipToSegment(refEdge)
        val normalFace = refEdge.start.dot(colNormal)
        for (pt in arrayOf(clippedEdge.start, clippedEdge.end))
            if (pt.dot(colNormal) <= normalFace) collision.addContact(pt)
        return collision
    }

    // Diagonals method, convex only
//    private fun getCollisionInfo2(polygon: PolygonCollider2D): CollisionManifold? {
//        val diagonals = arrayOf(getDiagonals(), polygon.getDiagonals())
//        val edges = arrayOf(polygon.getEdges(), getEdges())
//        var overlaps = FloatArray(this.countVertices() + polygon.countVertices())
//
//        var normalsAccum = Vec2()
//        var rc: RaycastResult?
//        // Raycast each shape's diagonals against other shape's edges
//        for (i in diagonals.indices) {
//            for (diag in diagonals[i]) {
//                for (side in edges[i]) {
//                    rc = side.raycast(Ray2D(diag), diag.length)
//                    if (rc != null)
//                        normalsAccum = normalsAccum.add(rc.normal * rc.distance)
//                }
//            }
//        }
//        if (normalsAccum.lenSquared() == 0f)
//            return null
//
//        val depth = normalsAccum.len()
//        val result = CollisionManifold(this, polygon, normalsAccum / depth, depth)
//        result.addContact(this.center())
//        result.addContact(polygon.center())
//        return result
//    }

    // SAT Helper Class
    private class SATPolygon(val collider: PolygonCollider2D, other: PolygonCollider2D) {
        val edges = collider.getEdges()
        val normals = collider.getNormals()

        // Find axis with greatest separation (least negative overlap)
        val overlaps = FloatArray(normals.size) { collider.getSeparationOnAxis(other, normals[it]) }
        val minIdx = maxIndex(*overlaps)
        val minOverlap = overlaps[minIdx]
    }

}